CLI: Add "automigrate" command for configuration issues and migrations

shilman opened the pull request:
Issue: #16150 

## What I did

<img width="749" alt="-bash" src="https://user-images.githubusercontent.com/488689/136006857-a03948e4-11f7-4abd-8ed5-820fb950a8cc.png">

- [x] Add new `sb automigrate` command
- [x] Add fixes for `webpack5`, `cra5`
- [x] Add `automigrate` to `sb upgrade`
- [x] Update CLI documentation

## How to test

- [ ] See attached tests
- [ ] Run `path/to/lib/cli/bin/index.js fix` in a project

shilman committed:
Fix CSFFile for arbitrary components

shilman committed:
CSF tools: Add utility for reading/writing config files

shilman committed:
CLI: Add fix command

shilman committed:
CLI: Add webpack5 fix

shilman committed:
CLI: Add cra5 fix

shilman committed:
Add links

shilman committed:
Update yarn.lock

kylegach commented on lib/cli/src/fix/helpers/getStorybookConfiguration.ts:
```diff
@@ -0,0 +1,21 @@
+/*
+ * Lifted from chromatic-cli
+ *
+ * This is not exactly clever but it works most of the time
```
😄

kylegach commented on MIGRATION.md:
```diff
@@ -330,9 +358,9 @@ npm install webpack@5 --save-dev
 Storybook 6.3 supports Angular 12 out of the box when you install it fresh. However, if you're upgrading your project from a previous version, you'll need to do the following steps to force Storybook to use webpack 5 for building your project:
```
As long as we're touching this file...

Ee point people to the "Webpack 5 manager build" heading often, and the directions there are incomplete. To work, you need to _also_ follow the directions under the "Angular 12 upgrade" heading, regardless of which framework they're using.

Can we please re-arrange these sections to avoid that confusion? Can attempt myself, if you'd prefer.

shilman replied:
would love a hand on that, thanks so much! 🙏 

we probably also need to document webpack5/builders in the main documentation site, and start pointing people there, FWIW

shilman replied:
we can also do that ☝️ as part of this little fire drill

kylegach commented on lib/cli/src/fix/fixes/cra5.ts:
```diff
@@ -0,0 +1,61 @@
+        'webpack5 builder'
+      )}.
+
+      See also ${chalk.yellow(
```
```suggestion
      More info: ${chalk.yellow(
```

Feels slightly more descriptive. Love that you thought to include the link here!

mrmckeb replied:
It would be great to have a convention regardless of the language used.

kylegach commented on lib/cli/src/fix/fixes/webpack5.ts:
```diff
@@ -0,0 +1,122 @@
+
+    if (semver.lte(storybookCoerced, '6.3.0')) {
+      logger.warn(
+        'Detected SB 6.3 or below, please upgrade storybook if you want to use webpack5.'
```
To do that, they can just run (or, at least, attempt to run) `npx sb@next upgrade`, yeah? We should tell them that.

mrmckeb replied:
It would be good to add this info, yes.

kylegach commented on lib/cli/src/fix/fixes/webpack5.ts:
```diff
@@ -0,0 +1,122 @@
+        'webpack5 builder'
+      )} for you.
+
+      See also ${chalk.yellow(
```
```suggestion
      More info: ${chalk.yellow(
```

kylegach commented on lib/cli/src/fix/index.ts:
```diff
@@ -0,0 +1,53 @@
+              {
+                type: 'confirm',
+                name: 'fix',
+                message: `Do you want to run fix '${chalk.cyan(f.id)}' on your project?`,
```
```suggestion
                message: `Do you want to run the '${chalk.cyan(f.id)}' fix on your project?`,
```

Reads a little more naturally, imo.

kylegach commented on lib/cli/src/fix/index.ts:
```diff
@@ -0,0 +1,53 @@
+      } else {
+        logger.info(`Skipping the ${chalk.cyan(f.id)} fix.`);
+        logger.info();
+        logger.info(`If you change your mind, just run '${chalk.cyan('npx sb@next fix')}'`);
```
```suggestion
        logger.info(`If you change your mind, run '${chalk.cyan('npx sb@next fix')}'`);
```

kylegach replied:
Also, if running `fix` is the only reasonable way someone would see this message, we should add "again" to the end.

shilman replied:
We're also going to run this as part of `sb upgrade`, so I'll leave off "again". Good thinking tho!

kylegach reviewed (commented):
I'll save the final approval for someone else, but looking at it from a DX and language perspective, this looks great! Left some comments and suggestions for you, but no blockers in that regard. Nice work!

mrmckeb commented on lib/cli/src/fix/fixes/webpack5.ts:
```diff
@@ -0,0 +1,122 @@
+
+const logger = console;
+
+/**
+ * Is the user using webpack5 in their project?
+ *
+ * If the user is using a version of SB >= 6.3,
+ * prompt them to upgrade to webpack5.
+ *
+ * - Add manager-webpack5 builder-webpack5 as dev dependencies
+ * - Add core.builder = 'webpack5' to main.js
+ * - Add 'webpack5' as a project dependency
+ */
```
A quick side-note that `/**` comments are JSDoc/TSDoc, so this would become documentation for the below interface.

Instead you can use `/*` or just `//`.

shilman replied:
Good catch -- moved down to be on the exported object instead

mrmckeb commented on lib/csf-tools/src/ConfigFile.ts:
```diff
@@ -0,0 +1,228 @@
+};
+
+export class ConfigFile {
+  _ast: t.File;
+
+  _exports: Record<string, t.Expression> = {};
+
+  _exportsObject: t.ObjectExpression;
```
Could you use private fields here?
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Private_class_fields#syntax

mrmckeb commented on lib/csf-tools/src/ConfigFile.ts:
```diff
@@ -0,0 +1,228 @@
+  }
+
+  parse() {
+    // eslint-disable-next-line @typescript-eslint/no-this-alias
+    const self = this;
```
Why do you alias this here, out of interest?

shilman replied:
Because `this` is bound to something else inside the Babel traversal 🤦

shilman committed:
CLI: Tests and bugfixes for webpack5 fix

shilman committed:
CLI: Tests and bugfix for cra5 fix

shilman committed:
Fix review comments

shilman committed:
More review fixes

shilman committed:
Add angular12 fix

kylegach committed:
Update MIGRATION.md, re: webpack 5

- This area was a point of confusion
- Attempted to reduce confusion while maintaining existing headings (so that anchor links already shared don't break)

shilman committed:
Rename `sb fix` to `sb automigrate`

shilman committed:
CLI: Add automigrate to `sb upgrade`

shilman committed:
Added upgrading documentation

shilman committed:
Merge branch 'next' into pr/13142

kylegach committed:
Edit upgrading doc page

kylegach commented on docs/configure/upgrading.md:
````diff
@@ -0,0 +1,45 @@
+npx sb@next upgrade
+```
+
+This upgrades all your Storybook packages to the latest stable version, perform sanity checks of your package versions, and also check for [automigrations](#automigrate) to automatically update your configuration.
````
I just pushed a commit with edits to this page, but we should avoid using the term "sanity check": https://www.selfdefined.app/definitions/sanity-check/

shilman replied:
🙀 thank you! 🙀

kylegach commented on docs/configure/upgrading.md:
```diff
@@ -2,44 +2,44 @@
 ## Automigrate script
 
-Storybook upgrades are not the only thing that can break your Storybook: changes in the ecosystem are also present challenges. For example, lots of frameworks (Angular 12, Create React App v5, NextJS) have recently migrated from webpack4 to webpack5, so even if you don't upgrade your Storybook version, you might need to update your configuration accordingly. That's what Automigrate is for:
+Storybook upgrades are not the only thing to consider: changes in the ecosystem also present challenges. For example, lots of frameworks ([Angular 12](https://angular.io/guide/updating-to-version-12#breaking-changes-in-angular-version-12), [Create React App v5](https://github.com/facebook/create-react-app/pull/11201), [NextJS](https://nextjs.org/docs/upgrading#webpack-5)) have recently migrated f ...
```
I attempted to add links to the most helpful spot for each of these. Not sure if that's actually more helpful than just linking to the projects' homepages. 🤷‍♂️

shilman replied:
It's great. 💯

shilman committed:
Fix typo

shilman committed:
Merge branch 'next' into 16150-cra5-sb-upgrade

shilman committed:
Merge branch 'next' into 16150-cra5-sb-upgrade
